### PR TITLE
feat(answers): add `findAnswers`

### PR DESCRIPTION
**Pull request by eunjae-lee**
This PR adds `findAnswers` method.
In this way, we can apply facets to /answers call.

It doesn't emit `'result'` but returns a promise of result.
If we can replace the traditional hits with the hits from answers completely, then we should go for emitting `'result'`, but it's not the case at the moment. So I chose to return a promise.

Let me know what you think.

(I haven't added any tests because we haven't agreed on how to implement this in helper yet. Once decided here, I will add test.)

### CODE REVIEW

File: src/algoliasearch.helper.js [hunk 1]
```diff
@@ -248,6 +250,55 @@ AlgoliaSearchHelper.prototype.searchOnce = function(options, cb) {
   });
 };
 
+/**
+ * Start the search for answers with the parameters set in the state.
+ * This method returns a promise.
+ * @return {AlgoliaSearchHelper}
+ * @return {promise.<FacetSearchResult>} the answer results
+ */
+AlgoliaSearchHelper.prototype.searchForAnswers = function(attributesForPrediction, queryLanguages, nbHits) {
```
Comment (Haroenv) [inline src/algoliasearch.helper.js:259]: these arguments will be more flexible in an object I think

Comment (Haroenv) [inline src/algoliasearch.helper.js:259]: the method needs to be added to the type definition still

File: src/algoliasearch.helper.js [hunk 2]
```diff
@@ -248,6 +250,55 @@ AlgoliaSearchHelper.prototype.searchOnce = function(options, cb) {
...
+      'hitsPerPage',
+      'restrictSearchableAttributes',
+      'clickAnalytics',
+      'facets',
+      'highlightPostTag',
+      'highlightPreTag',
+      'maxValuesPerFacet',
+      'page',
+      'snippetEllipsisText',
+      'tagFilters'
```
Comment (Haroenv) [inline src/algoliasearch.helper.js:282]: what's the reason for omitting these parameters?

Comment (eunjae-lee) [reply src/algoliasearch.helper.js]: It was my bad putting things in a wrong object.

File: src/algoliasearch.helper.js [hunk 3]
```diff
@@ -248,6 +250,50 @@ AlgoliaSearchHelper.prototype.searchOnce = function(options, cb) {
...
+    {
+      attributesForPrediction: attributesForPrediction,
+      nbHits: nbHits
+    },
+    {
+      params: omit(requestBuilder._getHitsSearchParams(derivedState), [
+        'attributesToSnippet',
+        'hitsPerPage',
+        'restrictSearchableAttributes',
+        'snippetEllipsisText' // FIXME remove this line once the engine is fixed.
```
Comment (eunjae-lee) [inline src/algoliasearch.helper.js:276]: Unlike [the doc](https://www.algolia.com/doc/rest-api/answers/#method-param-params), `snippetEllipsisText` is throwing an error. I've reported it to the team, but for now let's have it here.

File: src/algoliasearch.helper.js (hunk 1 above)
Comment (eunjae-lee) [reply src/algoliasearch.helper.js]: https://github.com/algolia/algoliasearch-helper-js/pull/804/commits/e1c783f9ce843d3279b12e8ce256e6f2794073ce

Comment (eunjae-lee) [reply src/algoliasearch.helper.js]: https://github.com/algolia/algoliasearch-helper-js/pull/804/commits/de96b4a734e2d6d5f7c0fb3fcb4d84debf7cb350
not sure how far I should go with the typing.

File: index.d.ts [hunk 4]
```diff
@@ -171,6 +171,28 @@ declare namespace algoliasearchHelper {
...
+     * @param {string[]} options.attributesForPrediction - Attributes to use for predictions. If empty, `searchableAttributes` is used instead.
+     * @param {string[]} options.queryLanguages - The languages in the query. Currently only supports ['en'].
+     * @param {number} options.nbHits - Maximum number of answers to retrieve from the Answers Engine. Cannot be greater than 1000.
+     */
+    findAnswers(options: {
+      attributesForPrediction: string[];
+      queryLanguages: string[];
+      nbHits: number;
+    }): Promise<{
+      hits: Array<any & {
```
Comment (Haroenv) [inline index.d.ts:187]: can you not use the return type of findAnswers on the client?

Comment (Haroenv) [reply index.d.ts:187]: this method likely should be generic, like the client one

File: src/algoliasearch.helper.js [hunk 5]
```diff
@@ -248,6 +250,75 @@ AlgoliaSearchHelper.prototype.searchOnce = function(options, cb) {
...
+ * @type {object}
+ * @property {Answer} _answer the object describing why the hit was chosen
+ */
+
+/**
+ * @typedef AnswersResult
+ * @type {object}
+ * @property {AnswerHit[]} hits the answer hits
+ */
+
```
Comment (Haroenv) [inline src/algoliasearch.helper.js:272]: these docs aren't used as explanation while there's the definition file (only locally), so I don't think it's needed.

Are you allowed to omit the return type all together or does eslint not allow it?

File: src/algoliasearch.helper.js [hunk 6]
```diff
@@ -248,6 +250,75 @@ AlgoliaSearchHelper.prototype.searchOnce = function(options, cb) {
...
+ /**
+ * Start the search for answers with the parameters set in the state.
+ * This method returns a promise.
+ * @param {Object} options - the options for answers API call
+ * @param {string[]} options.attributesForPrediction - Attributes to use for predictions. If empty, `searchableAttributes` is used instead.
+ * @param {string[]} options.queryLanguages - The languages in the query. Currently only supports ['en'].
+ * @param {number} options.nbHits - Maximum number of answers to retrieve from the Answers Engine. Cannot be greater than 1000.
+ *
+ * @return {AlgoliaSearchHelper}
```
Comment (Haroenv) [inline src/algoliasearch.helper.js:281]:
```suggestion
```

this line isn't right, you can't chain for now

File: src/algoliasearch.helper.js [hunk 7]
```diff
@@ -248,6 +250,75 @@ AlgoliaSearchHelper.prototype.searchOnce = function(options, cb) {
...
+    typeof this.client.initIndex !== 'function'
+  ) {
+    throw new Error(
+      'search for answers was called, but this client does not have a function client.initIndex'
+    );
+  }
+  var index = this.client.initIndex(derivedState.index);
+  if (typeof index.findAnswers !== 'function') {
+    throw new Error(
+      'search for answers was called, but this client does not have a function client.initIndex(index).findAnswers'
```
Comment (Haroenv) [inline src/algoliasearch.helper.js:316]: this is helpful, but the second error message gives pretty much the same information. Could it be consolidated to save space? How do we do it in searchForFacetValues?

Comment (eunjae-lee) [reply src/algoliasearch.helper.js]: Unlike SFFV, the existence of `findAnswers` can be checked only after creating `index` here. So I couldn't merge them in a single conditional block.
https://github.com/algolia/algoliasearch-helper-js/pull/804/commits/b6d1f57923d32e198265f3dd6cdfdbc75f0ec387

File: index.d.ts (hunk 4 above)
Comment (eunjae-lee) [reply index.d.ts]: I've updated the dependency and now it's using the type from the client:
https://github.com/algolia/algoliasearch-helper-js/pull/804/commits/da8af1338fe8038ef29b3786be0a0c789d6a2daf

File: src/algoliasearch.helper.js (hunk 5 above)
Comment (eunjae-lee) [reply src/algoliasearch.helper.js]: What do you think? https://github.com/algolia/algoliasearch-helper-js/pull/804/commits/5aa1206d303e8db790cf4fb6efa4070c3664a82d

File: index.d.ts [hunk 8]
```diff
@@ -171,6 +172,20 @@ declare namespace algoliasearchHelper {
...
+     * @param {Object} options - the options for answers API call
+     * @param {string[]} options.attributesForPrediction - Attributes to use for predictions. If empty, `searchableAttributes` is used instead.
+     * @param {string[]} options.queryLanguages - The languages in the query. Currently only supports ['en'].
+     * @param {number} options.nbHits - Maximum number of answers to retrieve from the Answers Engine. Cannot be greater than 1000.
+     */
+    findAnswers<TObject>(options: {
+      attributesForPrediction: string[];
+      queryLanguages: string[];
+      nbHits: number;
+    }): FindAnswersResponse<TObject>;
```
Comment (Haroenv) [inline index.d.ts:187]:
```suggestion
    }): Promise<FindAnswersResponse<TObject>>;
```

**Review by Haroenv — APPROVED** [review]
LGTM!

**Pull request merged**